File: app/src/main/java/com/dss886/dotaautochess/feature/match/manager/MatchManager.java
package com.dss886.dotaautochess.feature.match.manager;

import com.dss886.dotaautochess.data.Hero;
import com.dss886.dotaautochess.data.Profession;
import com.dss886.dotaautochess.data.Species;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.core.util.Pair;

/**
 * Created by dss886 on 2019/1/27.
 */
public class MatchManager {

    private static MatchManager instance;

    public static MatchManager inst() {
        if (instance == null) {
            synchronized (MatchManager.class) {
                if (instance == null) {
                    instance = new MatchManager();
                }
            }
        }
        return instance;
    }

    @NonNull
    private final ArrayList<Hero> mHeroList = new ArrayList<>();
    @NonNull
    private final List<MatchChangeListener> mChangeListeners = new ArrayList<>();

    private MatchManager() {
        for (int i = 0; i < 8; i++) {
            mHeroList.add(Hero.values()[i]);
        }
    }

    public void registerChangeListener(MatchChangeListener listener) {
        mChangeListeners.add(listener);
    }

    public void unregisterChangeListener(MatchChangeListener listener) {
        mChangeListeners.remove(listener);
    }

    private void notifyListeners() {
        for (MatchChangeListener listener : mChangeListeners) {
            listener.onHeroesChanged();
        }
    }

    public void addHero(Hero hero) {
        mHeroList.add(hero);
        notifyListeners();
    }

    public void removeHero(Hero hero) {
        mHeroList.remove(hero);
        notifyListeners();
    }

    @NonNull
    public List<Hero> getHeroList() {
        //noinspection unchecked
        return (List<Hero>) mHeroList.clone();
    }

    public List<Pair<Species, Integer>> getSpeciesList() {
        List<Species> speciesList = new ArrayList<>();
        List<Integer> countList = new ArrayList<>();
        for (Hero hero : mHeroList) {
            for (Species species : hero.speciesList) {
                int index = speciesList.indexOf(species);
                if (index <= 0) {
                    speciesList.add(species);
                    countList.add(1);
                } else {
                    countList.set(index, countList.get(index) + 1);
                }
            }
        }
        List<Pair<Species, Integer>> result = new ArrayList<>();
        for (int i = 0; i < speciesList.size(); i++) {
            result.add(new Pair<>(speciesList.get(i), countList.get(i)));
        }
        Collections.sort(result, new BuffComparator());
        return result;
    }

    public List<Pair<Profession, Integer>> getProfessionList() {
        List<Profession> professionList = new ArrayList<>();
        List<Integer> countList = new ArrayList<>();
        for (Hero hero : mHeroList) {
            int index = professionList.indexOf(hero.profession);
            if (index <= 0) {
                professionList.add(hero.profession);
                countList.add(1);
            } else {
                countList.set(index, countList.get(index) + 1);
            }
        }
        List<Pair<Profession, Integer>> result = new ArrayList<>();
        for (int i = 0; i < professionList.size(); i++) {
            result.add(new Pair<>(professionList.get(i), countList.get(i)));
        }
        Collections.sort(result, new BuffComparator());
        return result;
    }

}

File: app/src/main/java/com/dss886/dotaautochess/feature/match/manager/MatchManager.kt
package com.dss886.dotaautochess.feature.match.manager

import android.content.Context
import com.dss886.dotaautochess.app.App
import com.dss886.dotaautochess.data.Hero
import com.dss886.dotaautochess.data.Profession
import com.dss886.dotaautochess.data.Species
import java.util.*

/**
 * Created by dss886 on 2019/1/27.
 */
object MatchManager {

    private const val KEY_NAME = "hero"
    private val mSp = App.inst.getSharedPreferences("match", Context.MODE_PRIVATE)
    private val mHeroList = ArrayList<Hero>()
    private val mChangeListeners = ArrayList<MatchChangeListener>()

    val size: Int
        get() = mHeroList.size

    @Suppress("UNCHECKED_CAST")
    val heroList: List<Hero>
        get() = mHeroList.clone() as List<Hero>

    val speciesList: List<Pair<Species, Int>>
        get() = mHeroList.flatMap { it.speciesList.asIterable() }
                .groupBy { it }
                .map { Pair(it.key, it.value.size) }
                .sortedWith(BuffComparator())

    val professionList: List<Pair<Profession, Int>>
        get() = mHeroList.map { it.profession }
                .groupBy { it }
                .map { Pair(it.key, it.value.size) }
                .sortedWith(BuffComparator())

    init {
        readData()
    }

    fun registerChangeListener(listener: MatchChangeListener) {
        mChangeListeners.add(listener)
    }

    fun unregisterChangeListener(listener: MatchChangeListener) {
        mChangeListeners.remove(listener)
    }

    private fun notifyListeners() {
        for (listener in mChangeListeners) {
            listener.onHeroesChanged()
        }
    }

    fun addHero(hero: Hero): Boolean {
        if (!mHeroList.contains(hero) && mHeroList.size < 10) {
            mHeroList.add(hero)
            notifyListeners()
            saveData()
            return true
        }
        return false
    }

    fun containHero(hero: Hero): Boolean {
        return mHeroList.contains(hero)
    }

    fun removeHero(hero: Hero) {
        mHeroList.remove(hero)
        notifyListeners()
        saveData()
    }

    fun clearAllHero() {
        mHeroList.clear()
        notifyListeners()
        saveData()
    }

    private fun saveData() {
        val string = mHeroList.joinToString(",") { it.desc }
        mSp.edit().putString(KEY_NAME, string).apply()
    }

    private fun readData() {
        mSp.getString(KEY_NAME, "")
                ?.split(",")
                ?.forEach { desc ->
                    Hero.values().find { it.desc == desc }
                            ?.apply { mHeroList.add(this) }
                }
    }

}
